Adds Azure Function PowerShell guide (#6526)
pnp/cli-microsoft365: docs/docs/user-guide/automation/azure-function-powershell.mdx

Reviewer: This guide is quite detailed, providing an in-depth explanation of the entire setup process. However, it might be an idea to split it into more focused files, each dedicated to a separate section, to make it more digestible. Here’s a suggestion:

- Overview and Prerequisites
- Setting up Azure resources
- Setting up a local project
- Creating Functions
- Granting permissions and deployment
- Testing the functions

Additionally, the Overview page could include a list of all the pages that you'll need to follow for the guide.

Author: In the future we're planning to add more guides on how to use CLI in automation tools, right? If we split it into multiple pages, won't it become messy and difficult to find your way around?

Reviewer: Then I would create a subcategory, something like "Azure Functions" which then holds all these sections. Could of course be a bit overkill 😄

Author: I'm not really a fan of splitting it up into multiple pages. Personally, I would find it less clear, especially when we have multiple guides. We can always look for a second opinion. If the majority thinks that splitting it up is better, I'm ok with that.

    Make sure your file looks like this (choose your own project name and version):

    ```json title="package.json"
    {
      "name": "azure-function-cli-microsoft-365",
      "version": "1.0.0"
    }
    ```

1. Install CLI for Microsoft 365 in the project.

    Run the following command in your terminal at the root of your project. This will install the CLI for Microsoft 365 in the project locally.

    ```sh
    npm i @pnp/cli-microsoft365
    ```

1. Update `profile.ps1` file.

    The `profile.ps1` file is a script that runs every time your function app starts up on a new instance.
    This behavior is similar to how PowerShell profiles work on your local machine but tailored for the Azure Functions environment.
    The primary purpose of `profile.ps1` is to allow you to run initialization code for your function app.
    This could include setting up environment variables, loading custom modules, or executing any startup tasks required before your functions can run.
    In the case of CLI for Microsoft 365, we need to make sure that the CLI is available in the context of the Azure Function.

>     Overwrite the content of the `profile.ps1` file with the following content:

Reviewer: Sounds a bit clearer 😄 

```suggestion
    Replace the content of the `profile.ps1` file with the following code:
```

      ```
    </details>

## Granting permissions

In order to run the Azure Function, you need to grant the necessary permissions to the Azure Function.
In this guide, we are using the Azure managed identity to authenticate with CLI for Microsoft 365.
This way, you don't have to store any credentials in your Azure Function.

:::note

There are many ways to authenticate with CLI for Microsoft 365. In this guide, we are using the Azure managed identity.
If you want to use a different authentication method, please refer to the [log in to Microsoft 365 documentation](../connecting-microsoft-365) or the [login command documentation](../../cmd/login.mdx).

:::

1. Enable managed identity.

    Navigate to the `Identity` tab of your Azure Function and enable the system-assigned managed identity.
    Copy the Object ID of the managed identity, as you will need it in the next step.

    ![Enable managed identity](../../images/azure-function-powershell-guide/azure-function-identity.png)

1. Grant permissions to the managed identity.

    In order to use CLI for Microsoft 365, you need to grant the necessary permissions to the managed identity.

>     In this guide, out demo code needs the following permissions: `Sites.FullControl.All`, `Group.Read.All` and `Mail.Send`.

Reviewer: ```suggestion
    In this guide, our demo code needs the following permissions: `Sites.FullControl.All`, `Group.Read.All` and `Mail.Send`.
```

    Let's grant them to the managed identity.

    ```powershell
    # This script assumes you are logged into CLI for Microsoft 365
    # Replace the value of appObjectId with the object ID of your managed identity
    m365 entra approleassignment add --appObjectId "173b0035-8041-40da-9237-abf6de3be4a4" --resource "SharePoint" --scopes "Sites.FullControl.All"
    m365 entra approleassignment add --appObjectId "173b0035-8041-40da-9237-abf6de3be4a4" --resource "Microsoft Graph" --scopes "Group.Read.All,Mail.Send"
    ```

## Deploying the project

This section will guide you through deploying your Azure Function to Azure.

1. Sign in to Azure on Visual Studio Code.

    In Visual Studio Code, open the command palette by pressing `Ctrl+Shift+P` (Windows/Linux) or `⌘+Shift+P` (macOS).
    Search for `Azure: Sign In` and follow the instructions to sign in to your Azure account.

    ![Sign in to Azure](../../images/azure-function-powershell-guide/azure-sign-in.png)

1. Using the Azure extension, deploy your Azure Function.

    Open the Azure tab in Visual Studio Code and select the subscription, resource group, and Function App you created earlier.

>     Right click your Function App and choose the `Deploy to Function App...` option.

Reviewer: ```suggestion
    Right-click your Function App and choose the `Deploy to Function App...` option.
```

    ![Deploy to Azure](../../images/azure-function-powershell-guide/azure-deploy.png)

    This will deploy the entire project with the CLI for Microsoft 365 package to your Azure Function.

## Testing

To verify that the deployment was successful, when you click the `Overview` tab of your Azure Function, you should see your functions: 

![Azure Function overview](../../images/azure-function-powershell-guide/function-overview.png)

### HTTP trigger

Open the `GetSitesFromHub` function and click on the `Get Function URL` button.

![Get function URL](../../images/azure-function-powershell-guide/function-url.png)

> Copy the URL and paste it in your browser or an API request tool.

Reviewer: ```suggestion
Copy the URL and paste it into your browser or an API request tool.
```
